code/utils/modular_curves.h

```diff
@@ -0,0 +1,530 @@
+	}
+
+	template<typename result_type>
+	static inline float number_to_float(const result_type& number) {
+		// if constexpr(std::is_same_v<std::decay_t<result_type>, fix>) // TODO: Make sure we can differentiate fixes from ints.
+		//	return f2fl(number);
+		// else
+		if constexpr(std::is_integral_v<std::decay_t<result_type>>)
+			return static_cast<float>(number);
+		else if constexpr(std::is_floating_point_v<std::decay_t<result_type>>)
+			return static_cast<float>(number);
+		else {
+			static_assert(!std::is_same_v<result_type, result_type>, "Tried to return non-numeric value");
+			return 0.f;
+		}
+	}
+
+  public:
+	template<int tuple_idx, typename input_type>
+	static inline float grab(const input_type& input) {
+		const auto& result = grab_internal<std::decay_t<decltype(grab_from_tuple<tuple_idx, input_type>(input).get())>, grabbers...>(grab_from_tuple<tuple_idx, input_type>(input));
+		if constexpr (is_optional_v<typename std::decay_t<decltype(result)>>) {
+			if (result.has_value())
+				return number_to_float(result->get());
+			else
+				return 1.0f;
+		}
+		else {
+			return number_to_float(result.get());
+		}
+	}
+};
+
+template<auto grabber_fnc>
+struct modular_curves_functional_input {
+  private:
+	template<int tuple_idx, typename input_type>
+	static inline auto grab_from_tuple(const input_type& input) {
+		if constexpr(tuple_idx < 0)
+			return std::cref(input);
+		else
+			return std::cref(std::get<tuple_idx>(input));
+	}
+
+  public:
+	template<int tuple_idx, typename input_type>
+	static inline float grab(const input_type& input) {
+		if constexpr (std::is_pointer_v<std::remove_reference_t<input_type>>){
+			return grabber_fnc(*grab_from_tuple<tuple_idx, input_type>(input));
+		}
+		else {
+			return grabber_fnc(grab_from_tuple<tuple_idx, input_type>(input));
+		}
+	}
+};
+
+enum class ModularCurvesMathOperators {
+	addition,
+	subtraction,
+	multiplication,
+	division,
+};
+
+template <typename first, typename second, auto operation>
+struct modular_curves_math_input {
+  public:
+	template <int tuple_idx, typename input_type>
+	static inline float grab(const input_type& input) {
+		float first_value = first::template grab<tuple_idx, input_type>(input);
+		float second_value = second::template grab<tuple_idx, input_type>(input);
+		static_assert(std::is_same_v<decltype(operation), ModularCurvesMathOperators>, "Operation type must be in ModularCurvesMathOperators!");
+		if constexpr (operation == ModularCurvesMathOperators::addition) {
+			return first_value + second_value;
+		} else if constexpr (operation == ModularCurvesMathOperators::subtraction) {
+			return first_value - second_value;
+		} else if constexpr (operation == ModularCurvesMathOperators::multiplication) {
+			return first_value * second_value;
+		} else if constexpr (operation == ModularCurvesMathOperators::division) {
+			if (second_value != 0.f)
+				return first_value / second_value;
+			else
+				return 1.f;
```

Comment:
So, definitely not NaN, because that will propagate to the final values.
Since this is an input, whether its 0 or 1 is just a matter of convention.
Most cases where the second value is 0 will be nonsensical anyways, but it must be set to a non-nan, non-crashing, non-infinite value.